lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs

```diff
@@ -278,16 +296,71 @@ instance Buildable (PartialTx era) where
         cardanoTxF :: Cardano.Tx era -> Builder
         cardanoTxF tx' = pretty $ pShow tx'
 
+data CoinSelection = forall k ktype. CoinSelection
+    { txLayer
+        :: TransactionLayer k ktype SealedTx
+        -- TODO: Replace with smaller and smaller parts of 'TransactionLayer'
+    , inputScriptLookup
+        :: Maybe ([(W.TxIn, W.TxOut)] -> [CA.Script KeyHash])
+    , inputScriptTemplate
+        :: Maybe ScriptTemplate
+
+
+    , canSpendFrom :: Address -> Bool
+    }
```

Comment:
Hi @Anviking! 

Just curious as to the purpose of the `CoinSelection` type. Is it a way to garther all the dependencies that `balanceTx` needs into a single record?

If so, would be `BalanceTxContext` a good alternative name?

Reply:
>  Is it a way to garther all the dependencies that balanceTx needs into a single record?

No 🤠

Reply:
> Is it a way to gather all the dependencies that balanceTx needs into a single record?

So it's not "*all* the dependencies", but rather the ones relating to how to select inputs from the `UTxO`. Notably this `CoinSelection` type is distinct from things like `PParams` and `ChangeAddressGen`. The latter because we can in theory use a Shelley wallet `UTxO` for inputs, but generate Byron-style change addresses.

I did actually consider combining it with the `UTxO` type into some kind of `data BalancingSource UTxO TxLayer/assumptions about utxo`, but I wasn't sure the name / overhead was good enough.

I'm open to naming suggestions, but I'm sceptical of `BalanceTxContext`, which sounds like something we could just throw all arguments into without meaningful boundaries.

Reply:
@Anviking thanks for you explanation, I appreciate it.

The issue I have with the term `CoinSelection` is that it's already overloaded elsewhere. If I see a type called `CoinSelection`, my first instinct is to think that it represents a selection (noun) of coins, or a process for selecting coins, which is obviously not what this type does.

I'd propose to choose a name that fits the intent, which is:
> dependencies ... relating to how to select inputs from the UTxO

How about:

- `InputSelectionContext`
- `UTxOSelectionContext`

I think the suffix `Context` is especially good here, as it conveys the idea that objects of this type provide the **_necessary context_** for balanceTx to know how to select inputs from a UTxO set.

Reply:
- `CoinSelector`? 🤔

Reply:
- `CoinSelectionContext`? 🤔 (I mean this is basically the same thing as `{Input, UTxO}`SelectionContext)

Reply:
We've gone with `UTxOAssumptions` now as part of the preceding #3811